### PR TITLE
Reduce kernel duplication

### LINKED ISSUE #447: Form with domain markers creates duplicated kernels
MWE:
```python
coord_element = VectorElement("Lagrange", interval, 1)
mesh = Mesh(coord_element)

element = FiniteElement("Lagrange", interval, 1)
V = FunctionSpace(mesh, element)

u = TrialFunction(V)
v = TestFunction(V)

a = inner(u, v) * dx((1,2))
```

Produces 2 identitical kernels:

```c
// Code for integral integral_8799b8ef6f5269ca9a672cec402eefb7d5222475

void tabulate_tensor_integral_8799b8ef6f5269ca9a672cec402eefb7d5222475(double* restrict A,
                                    const double* restrict w,
                                    const double* restrict c,
                                    const double* restrict coordinate_dofs,
                                    const int* restrict entity_local_index,
                                    const uint8_t* restrict quadrature_permutation)
{
    // Quadrature rules
    static const double weights_4a8[2] = { 0.5, 0.5 };
    // Precomputed values of basis functions and precomputations
    // FE* dimensions: [permutation][entities][points][dofs]
    static const double FE0_C0_Q4a8[1][1][2][2] =
        { { { { 0.7886751345948129, 0.2113248654051871 },
              { 0.2113248654051871, 0.7886751345948129 } } } };
    static const double FE1_C0_D1_Q4a8[1][1][1][2] = { { { { -1.0, 1.0 } } } };
    // Quadrature loop independent computations for quadrature rule 4a8
    const double J_c0 = coordinate_dofs[0] * FE1_C0_D1_Q4a8[0][0][0][0] + coordinate_dofs[3] * FE1_C0_D1_Q4a8[0][0][0][1];
    double sp_4a8[1];
    sp_4a8[0] = fabs(J_c0);
    for (int iq = 0; iq < 2; ++iq)
    {
        const double fw0 = sp_4a8[0] * weights_4a8[iq];
        double t0[2];
        for (int i = 0; i < 2; ++i)
            t0[i] = fw0 * FE0_C0_Q4a8[0][0][iq][i];
        for (int i = 0; i < 2; ++i)
            for (int j = 0; j < 2; ++j)
                A[2 * i + j] += FE0_C0_Q4a8[0][0][iq][j] * t0[i];
    }
}


void tabulate_tensor_integral_b1ad34778d813d876edcdf491953bf64df0b3699(double* restrict A,
                                    const double* restrict w,
                                    const double* restrict c,
                                    const double* restrict coordinate_dofs,
                                    const int* restrict entity_local_index,
                                    const uint8_t* restrict quadrature_permutation)
{
    // Quadrature rules
    static const double weights_4a8[2] = { 0.5, 0.5 };
    // Precomputed values of basis functions and precomputations
    // FE* dimensions: [permutation][entities][points][dofs]
    static const double FE0_C0_Q4a8[1][1][2][2] =
        { { { { 0.7886751345948129, 0.2113248654051871 },
              { 0.2113248654051871, 0.7886751345948129 } } } };
    static const double FE1_C0_D1_Q4a8[1][1][1][2] = { { { { -1.0, 1.0 } } } };
    // Quadrature loop independent computations for quadrature rule 4a8
    const double J_c0 = coordinate_dofs[0] * FE1_C0_D1_Q4a8[0][0][0][0] + coordinate_dofs[3] * FE1_C0_D1_Q4a8[0][0][0][1];
    double sp_4a8[1];
    sp_4a8[0] = fabs(J_c0);
    for (int iq = 0; iq < 2; ++iq)
    {
        const double fw0 = sp_4a8[0] * weights_4a8[iq];
        double t0[2];
        for (int i = 0; i < 2; ++i)
            t0[i] = fw0 * FE0_C0_Q4a8[0][0][iq][i];
        for (int i = 0; i < 2; ++i)
            for (int j = 0; j < 2; ++j)
                A[2 * i + j] += FE0_C0_Q4a8[0][0][iq][j] * t0[i];
    }
}
 
```

### PR DESCRIPTION
**Pull request by jorgensd**
Coupled with https://github.com/FEniCS/ufl/pull/92
Resolves #447

### CODE REVIEW

File: test/test_jit_expression.py [hunk 1]
```diff
@@ -174,12 +174,11 @@ def test_elimiate_zero_tables_tensor(compile_args):
 
     # Using same basix element for coordinate element and coefficient
     coeff_points = basix_c_e.points
-
     # Compile expression at interpolation points of second order Lagrange space
     b_el = basix.create_element(basix.ElementFamily.P, basix.cell.string_to_type(cell), 0, True)
     points = b_el.points
     obj, module, code = ffcx.codegeneration.jit.compile_expressions(
-        [(expr, points)], cffi_extra_compile_args=compile_args)
+        [(expr, points)])
```
Comment (jorgensd) [inline test/test_jit_expression.py:181]:
```suggestion
        [(expr, points)], cffi_extra_compile_args=compile_args)
```

File: ffcx/ir/representation.py [hunk 2]
```diff
@@ -79,8 +79,9 @@ def compute_ir(analysis, object_names, prefix, parameters, visualise):
     for fd_index, fd in enumerate(analysis.form_data):
         form_names[fd_index] = naming.form_name(fd.original_form, fd_index, prefix)
         for itg_index, itg_data in enumerate(fd.integral_data):
-            integral_names[(fd_index, itg_index)] = naming.integral_name(fd.original_form, itg_data.integral_type,
-                                                                         fd_index, itg_data.subdomain_id, prefix)
+            # Unique ID for integrals is 2**form_index * 3**integral_index
+            integral_names[(fd_index, itg_index)] = naming.integral_name(
+                fd.original_form, itg_data.integral_type, 2**fd_index * 3**itg_index, prefix)
```
Comment (michalhabera) [inline ffcx/ir/representation.py:84]: What is this `2**fd_index * 3**itg_index` for? Are you encoding multiple integers into one using prime factorisation?

Comment (jorgensd) [reply ffcx/ir/representation.py:84]: Yes, as we now only have a single kernel for integrals with multiple subdomain ids, we need to have a way of giving them a unique name based on the kernel and the location in the subdomain_ids.

Comment (michalhabera) [reply ffcx/ir/representation.py:84]: I would suggest to pass all integers that make up unique integral name and hash it all later in the naming.

Comment (jorgensd) [reply ffcx/ir/representation.py:84]: I'm not sure I follow. How would you do this?

Comment (michalhabera) [reply ffcx/ir/representation.py:84]:
```suggestion
                fd.original_form, itg_data.integral_type, fd_index, itg_index, prefix)
```
and
```python
def integral_name(original_form, integral_type, form_id, integral_id, prefix):
    sig = compute_signature([original_form], str((prefix, integral_type, form_id, integral_id)))
```

**Comment by jorgensd** [comment]
Superseeded by: https://github.com/FEniCS/ffcx/pull/589

**Pull request closed without merging**